### ISSUE TITLE
JsonSerializer reading from Utf8JsonReader

### ISSUE THREAD

**Issue by BrennanConroy** [dotnet/corefx#36717]
Currently if you want to get an object out of a `Utf8JsonReader` you need to first use the reader with a temporary `JsonDocument`, then pass the `JsonElement.GetRawText()` from that document into the `JsonSerializer`.

There should be a way to do this directly from the serializer and avoid the temporary `JsonDocument` and `GetRawText()` call.

```c#
var reader = new Utf8JsonReader(someBytes, isFinalBlock: true, state: default);
using var token = JsonDocument.ParseValue(ref reader);
var obj = JsonSerializer.Parse(token.RootElement.GetRawText(), someType);
```

The requested API would look something like:
```c#
var reader = new Utf8JsonReader(someBytes, isFinalBlock: true, state: default);
var obj = JsonSerializer.Parse(ref reader, someType);
```

I did an perf test with the current approach vs. something similar to the proposed change

#### TLDR ####
Using something similar to `JsonSerializer.Parse(ref reader, someType);` could improve Op/s from `22,587` to `59,662` and allocations from `50.28KB` to `20.23KB` in the scenario I was testing.

Current Approach:

Method |          Input |    HubProtocol |     Mean |     Error |   StdDev |     Op/s |  Gen 0 | Allocated |
---- | - | - | - | - | - | - | - | -
  ReadSingleMessage | LargeArguments |           Json | 44.27 us | 1.3590 us | 4.007 us | 22,587.6 | 1.4648 |  50.28 KB 
  ReadSingleMessage | LargeArguments | NewtonsoftJson | 51.58 us | 1.0247 us | 2.552 us | 19,386.1 | 0.5493 |  20.68 KB 

Using the original `someBytes` and `reader.BytesConsumed` to strategically slice data off and pass it directly to `JsonSerializer.Parse(...)`:

Method |          Input |    HubProtocol |     Mean |     Error |   StdDev |     Op/s |  Gen 0 | Allocated |
---- | - | - | - | - | - | - | - | -
  ReadSingleMessage | LargeArguments |           Json | 16.76 us | 0.4949 us | 1.459 us | 59,662.9 | 0.5493 |  20.23 KB |
  ReadSingleMessage | LargeArguments | NewtonsoftJson | 51.86 us | 1.0326 us | 1.725 us | 19,282.0 | 0.5493 |  20.68 KB |

**Comment by ahsonkhan** [comment]
NewtonsoftJson 51.58 -> 39.31. Why did that one change?

**Comment by BrennanConroy** [comment]
I'm not sure, that one confused me too. I ran it twice and got the same result.

Maybe benchmark dotnet isn't as sandboxed as it could be
________________________________

**Comment by stephan-tolksdorf** [comment]
Such a `Parse` overload would allow to e.g. deserialize a `params` object in a JSON RPC request, without having to read (and validate) the JSON representation of the params object twice.

**Comment by rynowak** [comment]
We have a use case for this when writing custom converters. Frequently you want to use the JsonSerializer to read a value from the writer so that the serializer can do the right thing.

**Comment by steveharter** [comment]
The `ReadValue` terminology -- does just `Read` work? We have `ReadAsync` already and that would be consistent.

Design notes:
- The buffer in the reader must contain all of the data necessary to return the value - i.e. the reader shouldn't return `false` to request more data.
- If there is extra data in the buffer after deserialization that is OK - the deserializer should not throw. This is different than the normal Parse methods which don't allow additional data. This also requires the serializer to track the state, or inspect the reader's internal state to determine the current json depth.

**Comment by ahsonkhan** [comment]
> The `ReadValue` terminology -- does just `Read` work? We have `ReadAsync` already and that would be consistent.

The `ReadAsync` tries to read as much as possible. `ReadValue` would be consistent with `JsonDocument.ParseValue` which reads just one value (whether it be an object/array or primitive).

**Comment by terrajobst** [comment]
Approved because we approved the combined one #36714.

**Comment by BrennanConroy** [comment]
Did a before and after for this change:
Before:

Method |          Input |    HubProtocol |     Mean |     Error |   StdDev |     Op/s |  Gen 0 |  Gen 1 | Allocated |
------------------ |--------------- |--------------- |---------:|----------:|---------:|---------:|-------:|-------:|----------:|
 ReadSingleMessage | LargeArguments |           Json | 21.11 us | 1.6846 us | 4.967 us | 47,381.8 | 2.0752 | 0.0610 |  50.29 KB |
 ReadSingleMessage | LargeArguments | NewtonsoftJson | 19.43 us | 0.4236 us | 1.236 us | 51,469.2 | 0.7019 |      - |  20.68 KB |

After:

Method |          Input |    HubProtocol |      Mean |     Error |   StdDev |      Op/s |  Gen 0 | Allocated |
------------------ |--------------- |--------------- |----------:|----------:|---------:|----------:|-------:|----------:|
 ReadSingleMessage | LargeArguments |           Json |  9.199 us | 0.5671 us | 1.672 us | 108,709.4 | 0.6714 |  20.15 KB |
 ReadSingleMessage | LargeArguments | NewtonsoftJson | 19.627 us | 0.3924 us | 1.126 us |  50,950.9 | 0.7019 |  20.68 KB |

With custom slicing code (near optimal perf):

Method |          Input |    HubProtocol |      Mean |     Error |   StdDev |      Op/s |  Gen 0 | Allocated |
------------------ |--------------- |--------------- |----------:|----------:|---------:|----------:|-------:|----------:|
 ReadSingleMessage | LargeArguments |           Json |  8.244 us | 0.5701 us | 1.672 us | 121,294.4 | 0.6866 |  20.23 KB |
 ReadSingleMessage | LargeArguments | NewtonsoftJson | 20.251 us | 0.3925 us | 1.088 us |  49,381.2 | 0.7019 |  20.68 KB |
